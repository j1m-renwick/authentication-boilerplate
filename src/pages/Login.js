import React, {useState} from "react";
import {Button, Card, Error, Form, Input, Logo} from "../style/Styles";
import logoImg from "../images/logo.jpg";
import {useStoreActions} from 'easy-peasy';
import {useHistory} from "react-router-dom";
import {postRequest} from "../api/util";

function Login() {

    //pushing to the history is the same as a router Redirect component
    const history = useHistory();

    const setLoggedIn = useStoreActions(actions => actions.setLoggedIn);
    const setUsername = useStoreActions(actions => actions.setUsername);

    const [isError, setIsError] = useState(false);
    const [userName, setUserName] = useState("");
    const [password, setPassword] = useState("");

    function postLogin() {
        console.log("sending login request...")
        postRequest('/auth/login', {"username": userName, "password": password})
            .then(res => {
                // token cookie is saved by the browser here - any future requests to
                // the domain will include it by default.
                let success = res.status === 200
                setLoggedIn(success);
                setIsError(!success);
                if(success) {
                    setUsername(userName);
                    history.push("/admin");
                }
            }).catch(res => {
            setLoggedIn(false);
            setIsError(true);
        })
    }

    return (
        <Card>
            <h1>Login Page</h1>
            <Logo src={logoImg}/>
            <Form>
                <Input
                    type="username"
                    value={userName}
                    onChange={e => {
                        setUserName(e.target.value);
                    }}
                    placeholder="email"
                />
                <Input
                    type="password"
                    value={password}
                    onChange={e => {
                        setPassword(e.target.value);
                    }}
                    placeholder="password"
                />
                <Button onClick={postLogin}>Sign In</Button>
            </Form>
            {isError && <Error>The username or password provided was incorrect</Error>}
        </Card>
    );
}

export default Login;