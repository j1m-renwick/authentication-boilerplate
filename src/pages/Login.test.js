import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Login from "./Login";
import 'jest-styled-components';

// mocking out the useStoreActions easy-peasy hook
const mockSetState = jest.fn();

jest.mock('easy-peasy', () => ({
  useStoreActions: initial => mockSetState
}));

test('component renders as expected', () => {
  const component = render(<Login />).baseElement;
  expect(component).toMatchSnapshot();
});

test('Entering text in the inputs should show in the component', () => {
  const component = render(<Login />);
  const usernameInput = component.getByPlaceholderText("email");
  const passwordInput = component.getByPlaceholderText("password");
  fireEvent.change(usernameInput, { target: { value: '1two3' } });
  expect(usernameInput.value).toBe('1two3');
  fireEvent.change(passwordInput, { target: { value: '4five6' } });
  expect(passwordInput.value).toBe('4five6');
})
