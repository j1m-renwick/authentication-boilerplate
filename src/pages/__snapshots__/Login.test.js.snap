// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders as expected 1`] = `
.c0 {
  box-sizing: border-box;
  max-width: 50%;
  margin: 0 auto;
  padding: 0 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c3 {
  padding: 1rem;
  border: 1px solid #999;
  margin-bottom: 1rem;
  font-size: 0.8rem;
}

.c4 {
  background: linear-gradient(to bottom,#6371c7,#5563c1);
  border-color: #3f4eae;
  border-radius: 3px;
  padding: 1rem;
  color: white;
  font-weight: 700;
  width: 100%;
  margin-bottom: 1rem;
  font-size: 0.8rem;
}

.c1 {
  width: 80%;
  margin-bottom: 1rem;
}

<body>
  <div>
    <div
      class="c0"
    >
      <h1>
        Login Page
      </h1>
      <img
        class="c1"
        src="logo.jpg"
      />
      <div
        class="c2"
      >
        <input
          class="c3"
          placeholder="email"
          type="username"
          value=""
        />
        <input
          class="c3"
          placeholder="password"
          type="password"
          value=""
        />
        <button
          class="c4"
        >
          Sign In
        </button>
      </div>
    </div>
  </div>
</body>
`;
